File: internal/association/association_test.go
package association

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"testing"
)

type testCatalog struct {
	plans map[uint64]bool
	tasks map[uint64]uint64
}

func (c testCatalog) ValidatePlan(planID uint64) error {
	if !c.plans[planID] {
		return errors.New("not found")
	}
	return nil
}

func (c testCatalog) TaskPlan(taskID uint64) (uint64, error) {
	planID, ok := c.tasks[taskID]
	if !ok {
		return 0, errors.New("not found")
	}
	return planID, nil
}

func TestHostValidatesTargetsAndMintsMonotonicAssociations(t *testing.T) {
	root := t.TempDir()
	alias := filepath.Join(t.TempDir(), "project")
	if err := os.Symlink(root, alias); err != nil {
		t.Fatal(err)
	}
	host, err := NewHost(alias, 7, testCatalog{
		plans: map[uint64]bool{2: true},
		tasks: map[uint64]uint64{9: 2},
	})
	if err != nil {
		t.Fatal(err)
	}
	canonicalRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		t.Fatal(err)
	}
	first, err := host.Bind("opaque-live-id", PointerV1{
		Version: VersionV1, PlanID: 2, TaskID: 9,
	}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if first.ProjectRoot != canonicalRoot || first.Generation != 7 ||
		first.LiveID != "opaque-live-id" || first.Revision != 1 ||
		first.Target.PlanID != 2 || first.Target.TaskID != 9 {
		t.Fatalf("first association = %#v", first)
	}
	second, err := host.Bind("opaque-live-id", PointerV1{
		Version: VersionV1, PlanID: 2,
	}, &first)
	if err != nil {
		t.Fatal(err)
	}
	if second.Revision != 2 || second.Target.PlanID != 2 || second.Target.TaskID != 0 {
		t.Fatalf("second association = %#v", second)
	}
}

func TestHostRejectsUnsupportedStaleAndMismatchedAssociations(t *testing.T) {
	root := t.TempDir()
	host, err := NewHost(root, 3, testCatalog{
		plans: map[uint64]bool{1: true, 2: true},
		tasks: map[uint64]uint64{8: 2},
	})
	if err != nil {
		t.Fatal(err)
	}
	tests := []struct {
		name    string
		pointer PointerV1
		want    error
	}{
		{"unsupported version", PointerV1{Version: 2}, ErrUnsupportedVersion},
		{"task without plan", PointerV1{Version: 1, TaskID: 8}, ErrInvalidTarget},
		{"missing plan", PointerV1{Version: 1, PlanID: 99}, ErrInvalidTarget},
		{"missing task", PointerV1{Version: 1, PlanID: 2, TaskID: 99}, ErrInvalidTarget},
		{"mismatched task", PointerV1{Version: 1, PlanID: 1, TaskID: 8}, ErrInvalidTarget},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			if _, err := host.Bind("live", test.pointer, nil); !errors.Is(err, test.want) {
				t.Fatalf("Bind = %v, want %v", err, test.want)
			}
		})
	}

	prior, err := host.Bind("live", PointerV1{Version: 1}, nil)
	if err != nil {
		t.Fatal(err)
	}
	prior.Generation = 2
	if _, err := host.Bind("live", PointerV1{Version: 1}, &prior); !errors.Is(err, ErrStaleAssociation) {
		t.Fatalf("stale Bind = %v, want ErrStaleAssociation", err)
	}
	if _, err := NewHost(root, 0, nil); !errors.Is(err, ErrStaleAssociation) {
		t.Fatalf("zero-generation NewHost = %v, want ErrStaleAssociation", err)
	}
}

func TestAssociationJSONContainsOnlyLiveContextMetadata(t *testing.T) {
	encoded, err := json.Marshal(AssociationV1{
		Version:     VersionV1,
		ProjectRoot: "/project",
		Generation:  3,
		LiveID:      "opaque-live-id",
		Target:      TargetV1{PlanID: 2, TaskID: 9},
		Revision:    4,
	})
	if err != nil {
		t.Fatal(err)
	}
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(encoded, &fields); err != nil {
		t.Fatal(err)
	}
	for _, field := range []string{
		"version", "projectRoot", "generation", "liveId", "target", "revision",
	} {
		if _, ok := fields[field]; !ok {
			t.Fatalf("association JSON missing %q: %s", field, encoded)
		}
		delete(fields, field)
	}
	if len(fields) != 0 {
		t.Fatalf("association JSON exposes unexpected fields: %v", fields)
	}
	var target map[string]json.RawMessage
	if err := json.Unmarshal(json.RawMessage(`{"planId":2,"taskId":9}`), &target); err != nil {
		t.Fatal(err)
	}
	var actualTarget map[string]json.RawMessage
	associationJSON := struct {
		Target json.RawMessage `json:"target"`
	}{}
	if err := json.Unmarshal(encoded, &associationJSON); err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(associationJSON.Target, &actualTarget); err != nil {
		t.Fatal(err)
	}
	if len(actualTarget) != len(target) || actualTarget["planId"] == nil || actualTarget["taskId"] == nil {
		t.Fatalf("association target JSON = %s", associationJSON.Target)
	}
}

File: crates/ptrack-agent/src/association_test.rs
use std::collections::BTreeMap;
use std::fs;

use super::{
    ASSOCIATION_VERSION_V1, Association, AssociationCatalog, AssociationError, AssociationHost,
    AssociationPointer, AssociationTarget, association_generation, association_project_root,
    bind_association,
};
use crate::test_support::TempDirectory;

struct Catalog {
    plans: Vec<u64>,
    tasks: BTreeMap<u64, u64>,
}

impl AssociationCatalog for Catalog {
    fn validate_plan(&self, plan_id: u64) -> Result<(), String> {
        self.plans
            .contains(&plan_id)
            .then_some(())
            .ok_or_else(|| "not found".to_owned())
    }

    fn task_plan(&self, task_id: u64) -> Result<u64, String> {
        self.tasks
            .get(&task_id)
            .copied()
            .ok_or_else(|| "not found".to_owned())
    }
}

#[test]
fn host_validates_and_mints_monotonic_authority_free_associations() {
    let root = TempDirectory::new("ptrack-agent-association-root");
    let links = TempDirectory::new("ptrack-agent-association-link");
    let alias = links.path().join("project");
    #[cfg(unix)]
    std::os::unix::fs::symlink(root.path(), &alias).unwrap();
    #[cfg(windows)]
    std::os::windows::fs::symlink_dir(root.path(), &alias).unwrap();
    let catalog = Catalog {
        plans: vec![2],
        tasks: BTreeMap::from([(9, 2)]),
    };
    let host = AssociationHost::new(&alias, 7, Some(&catalog)).unwrap();
    let first = host
        .bind(
            "opaque-live-id",
            AssociationPointer {
                version: 1,
                plan_id: 2,
                task_id: 9,
            },
            None,
        )
        .unwrap();
    assert_eq!(
        first.project_root,
        fs::canonicalize(root.path()).unwrap().to_string_lossy()
    );
    assert_eq!(first.revision, 1);
    let second = host
        .bind(
            "opaque-live-id",
            AssociationPointer {
                version: 1,
                plan_id: 2,
                task_id: 0,
            },
            Some(&first),
        )
        .unwrap();
    assert_eq!(second.revision, 2);
    assert_eq!(
        second.target,
        AssociationTarget {
            plan_id: 2,
            task_id: 0
        }
    );
}

#[test]
fn host_rejects_unsupported_invalid_and_stale_associations() {
    let root = TempDirectory::new("ptrack-agent-association-errors");
    let catalog = Catalog {
        plans: vec![1, 2],
        tasks: BTreeMap::from([(8, 2)]),
    };
    let host = AssociationHost::new(root.path(), 3, Some(&catalog)).unwrap();
    let cases = [
        (
            AssociationPointer {
                version: 2,
                ..AssociationPointer::default()
            },
            "unsupported association version: 2",
        ),
        (
            AssociationPointer {
                version: 1,
                task_id: 8,
                ..AssociationPointer::default()
            },
            "invalid association target: task requires a plan",
        ),
        (
            AssociationPointer {
                version: 1,
                plan_id: 99,
                task_id: 0,
            },
            "invalid association target: plan #99: not found",
        ),
        (
            AssociationPointer {
                version: 1,
                plan_id: 2,
                task_id: 99,
            },
            "invalid association target: task #99: not found",
        ),
        (
            AssociationPointer {
                version: 1,
                plan_id: 1,
                task_id: 8,
            },
            "invalid association target: task #8 belongs to plan #2, not plan #1",
        ),
    ];
    for (pointer, expected) in cases {
        assert_eq!(
            host.bind("live", pointer, None).unwrap_err().to_string(),
            expected
        );
    }
    let mut previous = host
        .bind(
            "live",
            AssociationPointer {
                version: 1,
                ..AssociationPointer::default()
            },
            None,
        )
        .unwrap();
    previous.generation = 2;
    assert_eq!(
        host.bind(
            "live",
            AssociationPointer {
                version: 1,
                ..AssociationPointer::default()
            },
            Some(&previous)
        )
        .unwrap_err(),
        AssociationError::Stale(None)
    );
    assert_eq!(
        AssociationHost::new(root.path(), 0, None)
            .err()
            .unwrap()
            .to_string(),
        "stale association: workspace generation must be nonzero"
    );
}

#[test]
fn association_json_contains_only_live_context_metadata() {
    let encoded = serde_json::to_string(&Association {
        version: ASSOCIATION_VERSION_V1,
        project_root: "/project".to_owned(),
        generation: 3,
        live_id: "opaque-live-id".to_owned(),
        target: AssociationTarget {
            plan_id: 2,
            task_id: 9,
        },
        revision: 4,
    })
    .unwrap();
    assert_eq!(
        encoded,
        r#"{"version":1,"projectRoot":"/project","generation":3,"liveId":"opaque-live-id","target":{"planId":2,"taskId":9},"revision":4}"#
    );
    assert_eq!(
        serde_json::to_string(&AssociationPointer {
            version: 1,
            plan_id: 0,
            task_id: 0
        })
        .unwrap(),
        r#"{"version":1}"#
    );
}

#[test]
fn optional_host_entrypoints_fail_closed_and_getters_are_empty() {
    assert_eq!(association_project_root(None), std::path::Path::new(""));
    assert_eq!(association_generation(None), 0);
    assert_eq!(
        bind_association(
            None,
            "live",
            AssociationPointer {
                version: 1,
                ..AssociationPointer::default()
            },
            None,
        )
        .unwrap_err(),
        AssociationError::HostRequired
    );
}
